File: cmd/tube-designer/algorithm.go
package main

import (
	"fmt"
	"math"

	"github.com/nextmv-io/sdk/mip"
)

// optimizeCutting uses the Nextmv MIP API with the HiGHS back-end to find
// the optimal cutting plan.
func optimizeCutting(cuts []Cut, stockLen int, kerf float64) Solution {
	// --- 1. Generate all possible cutting patterns ---
	allCuts := make([]int, len(cuts))
	for i, c := range cuts {
		allCuts[i] = c.Length
	}
	patterns := generatePatterns(allCuts, stockLen, kerf)
	if len(patterns) == 0 {
		fmt.Println("Error: no valid cutting patterns could be generated.")
		return Solution{}
	}

	// --- 2. Build the MIP model ---
	model := mip.NewModel()
	model.Objective().SetMinimize() // minimise number of sticks

	patternVars := make([]mip.Var, len(patterns))
	for i := range patterns {
		patternVars[i] = model.NewInt(0, math.MaxInt64)
		model.Objective().NewTerm(1.0, patternVars[i]) // cost = 1 per stick
	}

	// Track demand for each cut length
	cutDemand := make(map[int]int)
	for _, cut := range cuts {
		cutDemand[cut.Length]++
	}

	// Add ≥-demand constraints for every cut length
	for cutLen, demand := range cutDemand {
		constr := model.NewConstraint(mip.GreaterThanOrEqual, float64(demand))
		for i, p := range patterns {
			count := 0
			for _, piece := range p {
				if piece == cutLen {
					count++
				}
			}
			if count > 0 {
				constr.NewTerm(float64(count), patternVars[i])
			}
		}
	}

	// --- 3. Solve with HiGHS ---
	solver, err := mip.NewSolver(mip.Highs, model)
	if err != nil {
		fmt.Printf("HiGHS solver init error: %v\n", err)
		return Solution{}
	}

	solution, err := solver.Solve(mip.SolveOptions{}) // ← only this line changed
	if err != nil {
		fmt.Printf("Solve error: %v\n", err)
		return Solution{}
	}
	if !solution.IsOptimal() {
		fmt.Println("Warning: solver did not prove optimality.")
	}

	// --- 4. Convert solver output to our domain structs ---
	result := Solution{}
	totalUsed := 0

	for i, p := range patterns {
		numSticks := int(math.Round(solution.Value(patternVars[i])))
		if numSticks == 0 {
			continue
		}

		cutSlice := make([]Cut, len(p))
		usedLen := 0
		for j, cl := range p {
			cutSlice[j] = Cut{Length: cl}
			usedLen += cl
		}
		if len(cutSlice) > 1 {
			usedLen += int(math.Round(float64(len(cutSlice)-1) * kerf))
		}

		for s := 0; s < numSticks; s++ {
			result.Sticks = append(result.Sticks, Stick{
				Cuts:     cutSlice,
				StockLen: stockLen,
				UsedLen:  usedLen,
				WasteLen: stockLen - usedLen,
			})
		}
		totalUsed += usedLen * numSticks
	}

	result.NumSticks = len(result.Sticks)
	result.TotalWaste = result.NumSticks*stockLen - totalUsed
	return result
}

// generatePatterns finds all possible ways a single stick can be cut.
func generatePatterns(availableCuts []int, stockLen int, kerf float64) [][]int {
	uniqueCutsMap := make(map[int]bool)
	for _, c := range availableCuts {
		uniqueCutsMap[c] = true
	}
	uniqueCuts := []int{}
	for c := range uniqueCutsMap {
		uniqueCuts = append(uniqueCuts, c)
	}

	var patterns [][]int
	var currentPattern []int
	var find func(int, int)
	kerfInt := int(math.Round(kerf))

	find = func(startIndex, remainingLen int) {
		if len(currentPattern) > 0 {
			pCopy := make([]int, len(currentPattern))
			copy(pCopy, currentPattern)
			patterns = append(patterns, pCopy)
		}

		for i := startIndex; i < len(uniqueCuts); i++ {
			cut := uniqueCuts[i]
			kerfCost := 0
			if len(currentPattern) > 0 {
				kerfCost = kerfInt
			}
			if remainingLen >= cut+kerfCost {
				currentPattern = append(currentPattern, cut)
				find(i, remainingLen-(cut+kerfCost))
				currentPattern = currentPattern[:len(currentPattern)-1]
			}
		}
	}

	find(0, stockLen)
	return patterns
}

File: algorithm.cpp
#include "algorithm.h"
#include <Highs.h>
#include <algorithm>
#include <cmath>
#include <iostream>
#include <set>
#include <unordered_map>

Solution optimizeCutting(const std::vector<Cut> &cuts, int stockLen,
                         double kerf) {
  // Step 1: Generate all possible cutting patterns
  std::vector<int> allCuts;
  for (const auto &cut : cuts) {
    allCuts.push_back(cut.length);
  }

  auto patterns = generatePatterns(allCuts, stockLen, kerf);
  if (patterns.empty()) {
    std::cerr << "Error: no valid cutting patterns could be generated."
              << std::endl;
    return Solution();
  }

  // Step 2: Build the MIP model using HiGHS
  Highs highs;
  HighsModel model;

  // Variables: one integer variable per pattern (number of sticks using that
  // pattern)
  model.lp_.num_col_ = patterns.size();
  model.lp_.col_cost_.resize(patterns.size(), 1.0); // Minimize number of sticks
  model.lp_.col_lower_.resize(patterns.size(), 0.0);
  model.lp_.col_upper_.resize(patterns.size(), kHighsInf);

  // All variables are integers
  model.lp_.integrality_.resize(patterns.size(), HighsVarType::kInteger);

  // Track demand for each cut length
  std::unordered_map<int, int> cutDemand;
  for (const auto &cut : cuts) {
    cutDemand[cut.length]++;
  }

  // Constraints: one per unique cut length
  model.lp_.num_row_ = cutDemand.size();
  model.lp_.row_lower_.resize(cutDemand.size());
  model.lp_.row_upper_.resize(cutDemand.size(), kHighsInf);

  // Build constraint matrix
  std::vector<int> Astart;
  std::vector<int> Aindex;
  std::vector<double> Avalue;

  Astart.push_back(0);

  // For each pattern (column)
  for (size_t i = 0; i < patterns.size(); i++) {
    // Count how many of each cut length this pattern provides
    std::unordered_map<int, int> patternCounts;
    for (int piece : patterns[i]) {
      patternCounts[piece]++;
    }

    // Add non-zero entries to the constraint matrix
    int row = 0;
    for (const auto &[cutLen, demand] : cutDemand) {
      if (patternCounts.count(cutLen) && patternCounts[cutLen] > 0) {
        Aindex.push_back(row);
        Avalue.push_back(patternCounts[cutLen]);
      }
      row++;
    }

    Astart.push_back(Aindex.size());
  }

  // Set constraint bounds (>= demand)
  int row = 0;
  for (const auto &[cutLen, demand] : cutDemand) {
    model.lp_.row_lower_[row] = demand;
    row++;
  }

  // Set the constraint matrix
  model.lp_.a_matrix_.format_ = MatrixFormat::kColwise;
  model.lp_.a_matrix_.start_ = Astart;
  model.lp_.a_matrix_.index_ = Aindex;
  model.lp_.a_matrix_.value_ = Avalue;

  // Set model sense (minimization)
  model.lp_.sense_ = ObjSense::kMinimize;

  // Step 3: Solve with HiGHS
  highs.passModel(model);

  // Configure solver options
  highs.setOptionValue("presolve", "on");
  highs.setOptionValue("mip_rel_gap", 0.01); // 1% optimality gap

  HighsStatus status = highs.run();

  if (status != HighsStatus::kOk) {
    std::cerr << "HiGHS solve error: "
              << highs.modelStatusToString(highs.getModelStatus()) << std::endl;
    return Solution();
  }

  const HighsSolution &solution = highs.getSolution();

  // Step 4: Convert solver output to our domain structs
  Solution result;
  int totalUsed = 0;

  for (size_t i = 0; i < patterns.size(); i++) {
    int numSticks = static_cast<int>(std::round(solution.col_value[i]));
    if (numSticks == 0)
      continue;

    // Create cuts for this pattern
    std::vector<Cut> cutSlice;
    int usedLen = 0;

    for (int cl : patterns[i]) {
      cutSlice.push_back(Cut(cl, 0));
      usedLen += cl;
    }

    // Add kerf
    if (cutSlice.size() > 1) {
      usedLen += static_cast<int>(std::round((cutSlice.size() - 1) * kerf));
    }

    // Create sticks using this pattern
    for (int s = 0; s < numSticks; s++) {
      Stick stick;
      stick.cuts = cutSlice;
      stick.stock_len = stockLen;
      stick.used_len = usedLen;
      stick.waste_len = stockLen - usedLen;
      result.sticks.push_back(stick);
    }

    totalUsed += usedLen * numSticks;
  }

  result.num_sticks = result.sticks.size();
  result.total_waste = result.num_sticks * stockLen - totalUsed;

  return result;
}

std::vector<std::vector<int>>
generatePatterns(const std::vector<int> &availableCuts, int stockLen,
                 double kerf) {
  // Get unique cut lengths
  std::set<int> uniqueCutsSet(availableCuts.begin(), availableCuts.end());
  std::vector<int> uniqueCuts(uniqueCutsSet.begin(), uniqueCutsSet.end());

  std::vector<std::vector<int>> patterns;
  std::vector<int> currentPattern;
  int kerfInt = static_cast<int>(std::round(kerf));

  // Recursive function to find all valid patterns
  std::function<void(int, int)> findPatterns = [&](int startIndex,
                                                   int remainingLen) {
    // Save current pattern if it has cuts
    if (!currentPattern.empty()) {
      patterns.push_back(currentPattern);
    }

    // Try adding more cuts
    for (int i = startIndex; i < static_cast<int>(uniqueCuts.size()); i++) {
      int cut = uniqueCuts[i];
      int kerfCost = currentPattern.empty() ? 0 : kerfInt;

      if (remainingLen >= cut + kerfCost) {
        currentPattern.push_back(cut);
        findPatterns(i, remainingLen - (cut + kerfCost));
        currentPattern.pop_back();
      }
    }
  };

  findPatterns(0, stockLen);

  return patterns;
}
